ipn/ipnlocal/local.go

```diff
@@ -2833,20 +2814,48 @@ func (b *LocalBackend) sendFileNotify() {
 	b.send(n)
 }
 
-// popBrowserAuthNow shuts down the data plane and sends an auth URL
-// to the connected frontend, if any.
-func (b *LocalBackend) popBrowserAuthNow() {
+// setAuthURL sets the authURL and triggers [LocalBackend.popBrowserAuthNow] if the URL has changed.
+// This method is called when a new authURL is received from the control plane, meaning that either a user
+// has started a new interactive login (e.g., by running `tailscale login` or clicking Login in the GUI),
+// or the control plane was unable to authenticate this node non-interactively (e.g., due to key expiration).
+// b.interact indicates whether an interactive login is in progress.
+func (b *LocalBackend) setAuthURL(url string) {
+	var popBrowser, keyExpired bool
+
 	b.mu.Lock()
-	url := b.authURL
-	expired := b.keyExpired
+	switch {
+	case url == "":
+		b.resetAuthURLLocked()
+	case b.authURL != url:
+		b.authURL = url
+		b.authURLTime = b.clock.Now()
+		// Always open the browser if the URL has changed.
+		// This includes the transition from no URL -> some URL.
+		popBrowser = true
+	default:
+		// Otherwise, only open it if the user explicitly requests interactive login.
+		popBrowser = b.interact
+	}
+	keyExpired = b.keyExpired
+	// Consume the StartLoginInteractive call, if any, that caused the control
+	// plane to send us this URL.
+	b.interact = false
 	b.mu.Unlock()
 
-	b.logf("popBrowserAuthNow: url=%v, key-expired=%v, seamless-key-renewal=%v", url != "", expired, b.seamlessRenewalEnabled())
+	if popBrowser {
+		b.popBrowserAuthNow(url, keyExpired)
+	}
+}
+
+// popBrowserAuthNow shuts down the data plane and sends an auth URL
+// to the connected frontend, if any.
```

Comment:
document keyExpired.

and looks like `url` can be empty? (based on `url != ""`). So document what that means.

Reply:
Documented the `keyExpired`. FWIW, the underlying logic of how we use it remains unchanged; we just extracted the
```go
b.mu.Lock()
keyExpired = b.keyExpired
b.mu.Unlock()
```
block out of the `popBrowserAuthNow` method to the call sites.

Also, without the changes in this PR applied, and regardless of them, I think we might not be calling `(*LocalBackend).stopEngineAndWait()` in all relevant scenarios. I've occasionally experienced issues over the past few months that I believe could be explained by this. So, we might want to review how `blockEngineUpdates`/`stopEngineAndWait` are being used, including in this method. I can take care of that, but I would prefer to handle it separately.

Reply:
Also, the `url` cannot be empty there. We check for it in `setAuthURL`.
I updated the docstring for `setAuthURL` to document it.